### DISCUSSION TITLE
Can´t train with a Custom dataset

### DISCUSSION THREAD

**Discussion by edgarcin95** [asteroid-team/asteroid#598] [Q&A] [answered]
I'm working on a project of Acoustic guitar separation I made a dataset with 205 songs with the WAVs of the Mix and the acoustic guitar.
I attach some screenshots of my code:
![Screenshot from 2022-02-14 15-28-21](https://user-images.githubusercontent.com/72329461/153949450-90f8ceb9-40a5-4baa-99e8-00a2ad52db58.png)
![Screenshot from 2022-02-14 15-28-48](https://user-images.githubusercontent.com/72329461/153949470-ed1312af-0150-49be-852e-d81c0e30051a.png)

When I train with this model:

	model = SuDORMRFImprovedNet(n_src=2, bn_chan=128, num_blocks=16, upsampling_depth=4,
				mask_act='relu', in_chan=None, fb_name='stft',
				kernel_size=21,	n_filters=512, stride=20, sample_rate=44100)

I get this:
![Screenshot from 2022-02-14 15-31-27](https://user-images.githubusercontent.com/72329461/153949960-245cb055-ddee-4a2c-923a-5538cd10430b.png)
![Screenshot from 2022-02-14 15-32-07](https://user-images.githubusercontent.com/72329461/153949978-d45c4be0-8ce0-4a64-9da5-74f9b562070e.png)

If I train with this model:

	model = DCUNet(architecture='DCUNet-10', stft_n_filters=24, stft_kernel_size=10,
			stft_stride=4, sample_rate=44100)

I get this:
![Screenshot from 2022-02-14 15-33-51](https://user-images.githubusercontent.com/72329461/153950250-f91bc6bf-e654-4d8b-8d27-bad3d7f4511d.png)
![Screenshot from 2022-02-14 15-34-14](https://user-images.githubusercontent.com/72329461/153950269-32651f28-7a05-41c5-93bb-b01b46a4bb85.png)

I don't know if my datasets or dataloaders are well implemented, or I need an specific model for my dataset. I want to use a model based on UNETs.
How can I fix my code?

**Comment by edgarcin95** [comment]
I think the problem is my dataloader or my dataset implementation. If I do this:

	from asteroid.data import LibriMix
	train_loader, val_loader = LibriMix.loaders_from_mini(task="sep_clean", batch_size=8)
	x, y = next(iter(train_loader))
	print(f'x shape={x.shape} dtype={x.dtype}')
	print(f'y shape={y.shape} dtype={y.dtype}')

I get this:
x shape=torch.Size([8, 24000]) dtype=torch.float32
y shape=torch.Size([8, 2, 24000]) dtype=torch.float32


But if I do it with my dataset:

	trDs = AGDS('../Datos/DB44-1kHz24bitsAlignd/', 'train', segs, 44100)
	trainLoader = DataLoader(trDs, batch_size=batchSize, shuffle=True,
				num_workers=nWorkers, drop_last=True)
	x, y = next(iter(trainLoader))
	print(f'x shape={x.shape} dtype={x.dtype}')
	print(f'y shape={y.shape} dtype={y.dtype}')

I get this:
x shape=torch.Size([16, 2, 264600]) dtype=torch.float32
y shape=torch.Size([16, 2, 264600]) dtype=torch.float32

The input tensors don't have the same shape. In my Dataloader, the X corresponds to the tensor of my Mix and the Y is the tensor of the Acoustic Guitar. What do the input tensor of the LibriMix Dataloader corresponds to? Do you have any examples to build Dataloaders properly?

**Comment by edgarcin95** [comment]
Or do you recommend to build my own model-architecture with these: https://asteroid.readthedocs.io/en/latest/package_reference/blocks.html ?

**Comment by JorisCos** [comment, marked as answer]
For your first message the first error that you have comes from a bug on our side I will correct it asap. The second error comes from your configuration `stft_nfilters=24` it should be a multiple of 32 so minimum value of  `stft_nfilters=64`. 
For your dataset, if I understand correctly you have songs and you want to extract the acoustic guitar out of it. The models that you are using are designed for mono not stereo. So make sure this is ok for you. Then your input `signalX` should be the mix with shape `batch, 1, time` and your sources `signalY` should be `batch, n_src,time` where `n_src` is the number of sources. To be clear if you have a song with someone singing with a guitar and you just want the guitar then `n_src=1`and `signalY` is the isolated guitar if you want the voice and the guitar then `n_src=2` and `signalY` is a stacked Tensor made of the isolated voice and the isolated guitar

**Comment by edgarcin95** [comment]
Thanks for the support, I later figured out that my dataset should be mono. Are there models for stereo?
For the SUDORMRF model I modified the parameters to this: `SuDORMRFImprovedNet(n_src=1, bn_chan=32, num_blocks=4, upsampling_depth=2, mask_act='relu', in_chan=None, fb_name='free', kernel_size=21, n_filters=256, stride=25, sample_rate=44100)` and it works, but only with a batch size of 16.

**Comment by sabir-debuger** [comment]
hi, @edgarcin95 can i run your recipes on google colab???

**Reply by edgarcin95** [reply]
Yes you can!

**Reply by edgarcin95** [reply]
Tell me about how it went please. Thanks.
